medmodels/data_synthesis/mtgan/modules/preprocessor.py

```diff
@@ -0,0 +1,595 @@
+                patient_index=patient_index,
+                time_windows=time_windows,
+                absolute_time_window_attribute=absolute_time_window_attribute,
+                time_window_attribute=time_window_attribute,
+                number_of_windows_attribute=number_of_windows_attribute,
+            )
+
+        preprocessing_attributes = PreprocessingAttributes(
+            first_admission_attribute=first_admission_attribute,
+            time_window_attribute=time_window_attribute,
+            concept_index_attribute=concept_index_attribute,
+            number_of_windows_attribute=number_of_windows_attribute,
+            absolute_time_window_attribute=absolute_time_window_attribute,
+        )
+
+        return medrecord, preprocessing_attributes
+
+    def preprocess(
+        self, medrecord: MedRecord
+    ) -> Tuple[MedRecord, Dict[int, NodeIndex], PreprocessingAttributes]:
+        """Preprocess the MedRecord.
+
+        - Remove uncommon concepts.
+        - Sample patients.
+        - Find the first admission for each patient.
+        - Remove admissions with less than min_codes_per_window codes.
+        - Find the relative times for each patient per concept.
+
+        Args:
+            medrecord (MedRecord): The MedRecord.
+
+        Returns:
+            Tuple[MedRecord, Dict[int, NodeIndex], PreprocessingAttributes]:
+                The preprocessed MedRecord, the index to concept dictionary (a
+                dictionary that connects the index of the concept with name of the
+                concept), and the preprocessing attributes' names in the preprocessed
+                MedRecord.
+
+        Raises:
+            ValueError: If no patients or no concepts are in the MedRecord groups with
+                the given group names before preprocessing. Also another ValueError
+                raised if none are left after preprocessing in either group. If no
+                edges have the given time attribute, also a ValueError is raised.
+
+        Note:
+            It requires the time attribute to be present in the edges of the MedRecord
+                and its values to be datetime objects.
+            This method will remove patient nodes with less than two unique time
+                windows.
+        """
+        if not medrecord.select_edges(edge().has_attribute(self.time_attribute)):
+            raise ValueError("No edges in the MedRecord with that time attribute")
+
+        # TODO: copy of a MedRecord instead of modifying the original one once it is implemented
+        concept_index_attribute = self._get_attribute_name(medrecord, "concept_index")
+        medrecord, _ = self._remove_uncommon_concepts(
+            medrecord,
+            minimum_number_ocurrences=self.hyperparameters[
+                "minimum_occurrences_concept"
+            ],
+            concept_index_attribute=concept_index_attribute,
+        )
+        medrecord = self._sample_patients(
```

Comment:
Afterwards we remove patients again, so we will probably never get the number of patients that is stated in the Preprocessor